### PR TITLE
Switch to go-winres to create Windows resources

### PR DESCRIPTION
**Pull request by crazy-max**
follow-up https://github.com/docker/cli/pull/3310 and https://github.com/moby/moby/pull/42872

**- What I did**

Replace windres with native Go implementation to be able to create a syso file which contains Microsoft Windows Version Information, an icon and the Windows Event Logging Manifest using [go-winres](https://github.com/tc-hib/go-winres).

**- How I did it**

Remove windres references and replace with go-winres. Also use `go:generate` for better integration and add legal copyright field.

Also add cross target to the GHA ci workflow.

**- How to verify it**

```console
docker buildx bake --set *.args.DOCKER_CROSSPLATFORMS=windows/amd64 cross
```

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thank you!!!

@tianon @TBBle @corhere FYI; you seem like candidates that are interested in (reviewing) this 😅 🤗

**Comment by crazy-max** [comment]
Need more work on windows build side: https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-43431/1/pipeline/282#step-283-log-343

```
[2022-03-26T18:31:09.692Z] At D:\gopath\src\github.com\docker\docker@tmp\durable-4df80df9\powershellWrapper.ps1:3 char:1
[2022-03-26T18:31:09.692Z] + & powershell -NoProfile -NonInteractive -ExecutionPolicy Bypass -Comm ...
[2022-03-26T18:31:09.692Z] + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2022-03-26T18:31:09.692Z]     + CategoryInfo          : NotSpecified: (cmd\dockerd\bin_windows_amd64.go:String) [], RemoteException
[2022-03-26T18:31:09.692Z]     + FullyQualifiedErrorId : NativeCommandError
[2022-03-26T18:31:09.692Z]  
[2022-03-26T18:31:09.692Z] cmd\dockerd\bin_windows_amd64.go:4: running "goversioninfo": exec: "goversioninfo": executable file not found in %PATH%
```

But cross is ok: https://github.com/moby/moby/runs/5704937248?check_suite_focus=true

**Comment by thaJeztah** [comment]
One failure; looks related (but could be due to how the test is run);

```
Stacktrace
=== RUN   TestVersion
    version_test.go:19: assertion failed: version.Version is ""
--- FAIL: TestVersion (0.03s)

**Comment by crazy-max** [comment]
> One failure; looks related (but could be due to how the test is run);
> 
> ```
> Stacktrace
> === RUN   TestVersion
>     version_test.go:19: assertion failed: version.Version is ""
> --- FAIL: TestVersion (0.03s)
> ```

Yes version string is not propagated the same way in the powershell script. Should be ok now.

File: Dockerfile [hunk 1]
```diff
@@ -346,34 +355,42 @@ ARG PRODUCT
 ENV PRODUCT=${PRODUCT}
 ARG DEFAULT_PRODUCT_LICENSE
 ENV DEFAULT_PRODUCT_LICENSE=${DEFAULT_PRODUCT_LICENSE}
+ARG COMPANY_NAME
+ENV COMPANY_NAME=${COMPANY_NAME}
```
Comment (thaJeztah) [inline Dockerfile:359]: Was thinking again about this one, and (also per our chats on slack), perhaps `PACKAGER_NAME` is a more descriptive name for this (also see https://github.com/docker/cli/pull/3500)

**Comment by crazy-max** [comment]
@thaJeztah Looks unrelated: https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-43431/18/pipeline/288#step-289-log-1466

```
[2022-03-27T09:44:02.016Z] === Failed
[2022-03-27T09:44:02.016Z] === FAIL: github.com/docker/docker/libnetwork/networkdb TestNetworkDBCRUDTableEntries (8.81s)
[2022-03-27T09:44:02.016Z]     networkdb_test.go:303: Entry existence verification test failed for node1(3ed9271ef286)
[2022-03-27T09:44:02.016Z] 2022/03/27 09:42:27 Closing DB instances...
```

**Comment by crazy-max** [comment]
@thaJeztah This one looks flaky too when the connection is already closed after the first dial hijack. Will fix that on buildkit: https://github.com/moby/moby/runs/5708867489?check_suite_focus=true#step:9:1746

```
=== CONT  TestClientGatewayIntegration/TestClientGatewayContainerSecurityMode/worker=dockerd/secmode=insecure
    build_test.go:1652: dockerd worker can not currently run this test due to missing features ()
    run.go:196: 
        	Error Trace:	run.go:196
        	            				panic.go:642
        	            				testing.go:858
        	            				testing.go:842
        	            				dockerd.go:202
        	            				build_test.go:1652
        	            				run.go:87
        	            				run.go:201
        	Error:      	Received unexpected error:
        	            	unexpected EOF
```

File: cmd/docker-proxy/bin_windows_386.go [hunk 2]
```diff
@@ -0,0 +1,8 @@
+//go:build windows && 386
+// +build windows,386
+
+//go:generate goversioninfo -o=../../cli/winresources/docker-proxy/resource.syso -icon=winresources/docker-proxy.ico -manifest=winresources/docker-proxy.exe.manifest ../../cli/winresources/docker-proxy/versioninfo.json
```
Comment (TBBle) [inline cmd/docker-proxy/bin_windows_386.go:4]: Would it be better to have the output resource.syso be `resource_<platform>_<arch>.syso`? Then it could perhaps be committed if feasible, and it definitely wouldn't be at-risk of doing builds for different platforms/arches in a row and forgetting to `go generate` between each one.

I thought the usual idiom was to commit the `go generate`-created artefacts to the repo, but I'm working from a _really_ limited set of inputs there. But in this case, I think I understand some of the inputs are going to vary at build-time, and we have the Makefile-based system already to trigger `go generate` on-demand.

File: cli/winresources/docker-proxy/winresources.go [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+//
+// These resources are used to provide:
+// * Version information
+// * An icon
+// * A Windows manifest declaring Windows version support
+//
+// The resource object files are generated when building with goversioninfo
+// in hack/make/.go-autogen and are located in cmd/docker-proxy/winresources.
+// This occurs automatically when you cross build against Windows OS.
+package winresources
```
Comment (TBBle) [inline cli/winresources/docker-proxy/winresources.go:11]: I wonder if instead of `cli/winresources/docker-proxy` it could be put in `cmd/docker-proxy/winresources`? That way it's closer to the thing it's specific to. In fact, it probably wouldn't need to be a subdirectory, what if the syso is output directly into `cmd/docker-proxy`? If as I suggested elsewhere it gets the right suffixes, then it won't interfere with any cross-builds etc either.

(Where `docker-proxy` means the same thing for `dockerd`.)

Actually, I just noticed that the comment _says_ it does what I'm suggesting, but the `go:generate` line is outputting into `cli/winresources` where this file currently lives. Was a restructuring done away from `cmd/docker-proxy/winresources` to `cli/winresources/docker-proxy` at some point?

File: cli/winresources/dockerd/winresources.go [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// Package winresources is used to embed Windows resources into dockerd.exe.
+//
+// These resources are used to provide:
+// * Version information
+// * An icon
+// * A Windows manifest declaring Windows version support
+//
+// The resource object files are generated when building with goversioninfo
+// in hack/make/.go-autogen and are located in cmd/docker-proxy/winresources.
```
Comment (TBBle) [inline cli/winresources/dockerd/winresources.go:9]: Nit: cut-and-paste error, `docker-proxy` should be `dockerd`.

File: cmd/docker-proxy/winresources/docker-proxy.exe.manifest [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+            <!-- Windows 10 -->
+            <supportedOS Id="{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}"/>
+            <!-- Windows 8.1 -->
+            <supportedOS Id="{1f676c76-80e1-4239-95bb-83d0f6d0da78}"/>
+            <!-- Windows Vista -->
+            <supportedOS Id="{e2011457-1546-43c5-a5fe-008deee3d3f0}"/> 
+            <!-- Windows 7 -->
+            <supportedOS Id="{35138b9a-5d96-4fbd-8e2d-a2440225f93a}"/>
+            <!-- Windows 8 -->
+            <supportedOS Id="{4a2f28e3-53b9-4441-ba9c-d69d4a4a6e38}"/>
```
Comment (TBBle) [inline cmd/docker-proxy/winresources/docker-proxy.exe.manifest:15]: Does it make sense to manifest for versions of Windows we don't support? I'm not sure what impact the manifest versions have for the UX, except that they tell `GetVersionEx` which lies to tell.

File: cmd/dockerd/bin_windows_arm64.go [hunk 6]
```diff
@@ -0,0 +1,8 @@
+//go:build windows && arm64
+// +build windows,arm64
+
+//go:generate goversioninfo -arm=true -64=true -o=../../cli/winresources/dockerd/resource.syso -icon=winresources/dockerd.ico -manifest=winresources/dockerd.exe.manifest ../../cli/winresources/dockerd/versioninfo.json
+
+package main
+
+import _ "github.com/docker/docker/cli/winresources/dockerd"
```
Comment (TBBle) [inline cmd/dockerd/bin_windows_arm64.go:8]: I'd suggest maybe `bin` is too generic a name for this file. What about `genversioninfo_windows_arm64.go` and friends? Something that gives a hint what it does?

File: hack/make.ps1 [hunk 7]
```diff
@@ -424,10 +425,17 @@ Try {
     # Run autogen if building binaries or running unit tests.
     if ($Client -or $Daemon -or $TestUnit) {
         Write-Host "INFO: Invoking autogen..."
-        Try { .\hack\make\.go-autogen.ps1 -CommitString $gitCommit -DockerVersion $dockerVersion -Platform "$env:PLATFORM" -Product "$env:PRODUCT" }
+        Try { .\hack\make\.go-autogen.ps1 -CommitString $gitCommit -DockerVersion $dockerVersion -Platform "$env:PLATFORM" -Product "$env:PRODUCT" -PackagerName "$env:PACKAGER_NAME" }
         Catch [Exception] { Throw $_ }
     }
```
Comment (TBBle) [inline hack/make.ps1:430]: I realise it's not changing here, but it seems like `.go-autogen.ps1` only knows how to autogen for `$Daemon` case? Is there any point in calling it for the `$Client` or `$TestUnit` cases?

**Review by TBBle — COMMENTED** [review]
This looks like a good solution, and I'm thinking it'll be worth stealing for hcsshim and hence containerd, since there was [a discussion just this week](https://github.com/containerd/containerd/pull/6697#discussion_r833850950) that the current manifesting approach there leaves a lot to be desired, since it relies on an in-tree .syso that someone created once upon a time.

I have some specific thoughts and ideas below, but since I haven't been involved in any of the work that's happened before now, it's possible I'm asking questions that have already been discussed and resolved elsewhere, so I apologise for the noise if that turns out to be the case.

Also, I'm not super-familiar with the in-container build system, so I haven't look carefully at that.

Edit: Linked to the discussion for context.

**Comment by thaJeztah** [comment]
> This looks like a good solution, and I'm thinking it'll be worth stealing for hcsshim and hence containerd, since there was a discussion just this week that the current manifesting approach there leaves a lot to be desired, since it relies on an in-tree .syso that someone created once upon a time.

Good suggestion; let me also ping @kevpar @dcantah @katiewasnothere

File: hack/make.ps1 (hunk 7 above)
Comment (crazy-max) [reply hack/make.ps1:430]: Good point, I was also wondering the same thing. As of now there is no `autogen` build tag and everything is handle through ldflags it should be fine to only autogen for `$Daemon`. I will check that.

**Comment by thaJeztah** [comment]
> I have some specific thoughts and ideas below, but since I haven't been involved in any of the work that's happened before now, it's possible I'm asking questions that have already been discussed and resolved elsewhere, so I apologise for the noise if that turns out to be the case.

A lot of code in this area was created during the initial implementation to make Docker work on Windows; some of that required working with many limitations (Golang not yet supporting the things we needed, Windows itself not yet supporting everything), so there will be a lot of areas that have (sometimes very dirty) hacks in place "to make it work". Some bits have been cleaned up, but others (as things tend to go in an evolving codebase) are left behind. Not all original contributors are still "around" (for example, John Howard, who did a major part of the first Windows implementation, has retired, and is now enjoying the good life 🤗)

TL;DR; any question/comment is welcome, and don't assume that everything you find was "by design" (it could well just have been "to make it work within the limitations we had at the time"

File: cmd/docker-proxy/winresources/docker-proxy.exe.manifest (hunk 5 above)
Comment (TBBle) [reply cmd/docker-proxy/winresources/docker-proxy.exe.manifest:15]: Also, nit for this file: No trailing newline.

File: cmd/docker-proxy/bin_windows_386.go (hunk 2 above)
Comment (crazy-max) [reply cmd/docker-proxy/bin_windows_386.go:4]: I don't think resources should be committed but generated at build time. But yes we could document that `go generate` should be run before build even if it's already the case in our Makefile.

File: cmd/dockerd/bin_windows_arm64.go (hunk 6 above)
Comment (crazy-max) [reply cmd/dockerd/bin_windows_arm64.go:8]: sure sgtm

File: cmd/docker-proxy/bin_windows_386.go (hunk 2 above)
Comment (TBBle) [reply cmd/docker-proxy/bin_windows_386.go:4]: It'd be nice if there was a way to make it _fail_ to build if `go generate` hadn't been run.

File: cli/winresources/docker-proxy/winresources.go (hunk 3 above)
Comment (crazy-max) [reply cli/winresources/docker-proxy/winresources.go:11]: We are mounting a tmpfs for `cli/winresources/docker-proxy` and `cli/winresources/dockerd` in our Dockerfile for generated resources (syso). If we move to `cmd/docker-proxy/winresources` or `cmd/dockerd/winresources` the stub package would be removed in there as well as manifest and icon. That why it's splitted that way. We could do that if we bind mount the working tree as `rw` instead of `ro` (default) but don't think it's a good idea.

File: cmd/docker-proxy/winresources/docker-proxy.exe.manifest [hunk 8]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">
+    <description>Docker Engine</description>
+    <compatibility xmlns="urn:schemas-microsoft-com:compatibility.v1"> 
+        <application> 
+            <!-- Windows 10 -->
+            <supportedOS Id="{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}"/>
```
Comment (thaJeztah) [inline cmd/docker-proxy/winresources/docker-proxy.exe.manifest:7]: https://docs.microsoft.com/en-us/windows/win32/sysinfo/targeting-your-application-at-windows-8-1
```suggestion
            <!-- Windows 10 and Windows 11 -->
            <supportedOS Id="{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}"/>
```

**Comment by thaJeztah** [comment]
let me post as a quick comment; in the maintainers meeting we were discussing if with this change, we would no longer require the "special" (only "compatible") versions of these in the dockerfile; https://github.com/moby/moby/blob/e69529c94a2624c46dd64e58fd634e1bb8a8bc9d/Dockerfile.windows#L227-L234

(would be nice if we would no longer need those, but of course, needs checking)


For CI, we could also speed up things by cross-compiling the windows binaries, and loading them into (a) windows machine(s) to run the tests; a large amount of time is consumed by just pulling the windows base image(s) and building the binary, and if cross-compiling works for that, we could save that time (and even more, if we can load the binaries in more machines to run the tests in parallel).

**Comment by thaJeztah** [comment]
Got a panic in buildkit; https://github.com/moby/moby/runs/5778480289?check_suite_focus=true

```
#52 DONE 0.1s
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xfca460]

goroutine 34 [running]:
github.com/moby/buildkit/util/progress/progressui.(*textMux).print(0xc000547e08, 0xc000547e80)
	/build/moby-buildx/.gopath/src/github.com/docker/buildx/vendor/github.com/moby/buildkit/util/progress/progressui/printer.go:284 +0x7c0
github.com/moby/buildkit/util/progress/progressui.DisplaySolveStatus(0x255d660, 0xc0001f0bc0, 0x0, 0x0, 0x0, 0x0, 0x25180c0, 0xc0000c6010, 0xc0000a2ae0, 0x0, ...)
	/build/moby-buildx/.gopath/src/github.com/docker/buildx/vendor/github.com/moby/buildkit/util/progress/progressui/display.go:83 +0x405
github.com/docker/buildx/util/progress.NewPrinter.func1(0x7fff9f21af64, 0x5, 0xc0001952e0, 0x2567118, 0xc0000c6010, 0x255d660, 0xc0001f0bc0, 0xc0000a2ae0, 0xc0004be500, 0xc0000a2c00)
	/build/moby-buildx/.gopath/src/github.com/docker/buildx/util/progress/printer.go:99 +0x132
created by github.com/docker/buildx/util/progress.NewPrinter
	/build/moby-buildx/.gopath/src/github.com/docker/buildx/util/progress/printer.go:87 +0x1e5
```

**Comment by crazy-max** [comment]
@thaJeztah https://github.com/moby/buildkit/issues/2762

**Comment by crazy-max** [comment]
> let me post as a quick comment; in the maintainers meeting we were discussing if with this change, we would no longer require the "special" (only "compatible") versions of these in the dockerfile;
> 
> https://github.com/moby/moby/blob/e69529c94a2624c46dd64e58fd634e1bb8a8bc9d/Dockerfile.windows#L227-L234
> 
> (would be nice if we would no longer need those, but of course, needs checking)
> 
> For CI, we could also speed up things by cross-compiling the windows binaries, and loading them into windows machine(s) to run the tests; a large amount of time is consumed by just pulling the windows base image(s) and building the binary, and if cross-compiling works for that, we could save that time (and even more, if we can load the binaries in more machines to run the tests in parallel).

I will check that in a follow-up if that sgty.

**Comment by thaJeztah** [comment]
> I will check that in a follow-up if that sgty.

for sure

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

File: cmd/docker-proxy/bin_windows_386.go (hunk 2 above)
Comment (corhere) [reply cmd/docker-proxy/bin_windows_386.go]: > I thought the usual idiom was to commit the go generate-created artefacts to the repo

That's what I thought as well. I went looking for [a primary source](https://go.dev/blog/generate) and found instead: (emphasis mine)
> Just keep in mind that it is for package authors, not clients, if only for the reason that the program it invokes might not be available on the target machine. Also, **if the containing package is intended for import by `go get`**, once the file is generated (and tested!) it must be checked into the source code repository to be available to clients.

So in the specific case of generated files used only as inputs for producing binary artifacts, it is fair game to require that they be generated at build time. Though perhaps committing .syso files for "stub" Windows resources containing only the manifest (and no version/copyright info) into the repo might be worth considering? Then people developing & testing on Windows could use `go build` to compile manifested executables, and only those building for release (i.e. packaging maintainers) need to have goversioninfo installed. Stub .syso files could be committed and `.gitignore`'d so that non-stub versions are not accidentally committed; `git add --force` can be used to update the stubs when necessary.

> It'd be nice if there was a way to make it _fail_ to build if `go generate` hadn't been run. But nothing comes immediately to mind.

Commit some kind of "poisoned" deliberately corrupt .syso which causes link to fail unless overwritten with a generated version? I can't imagine the resulting build error messages would be terribly informative, however. I think that out of all the options, committing stub .syso files would be the least-terrible choice as that would ensure that the Windows binaries will always be correctly manifested and therefore behave correctly at runtime, even if the version info is missing.

**Review by corhere — CHANGES_REQUESTED** [review]
The messages defined in event_messages.mc need to be included in the resources for dockerd.exe in order to support logging to the Windows Event log: 57aef3b49025aac6bb084491478eb461b14b9109. I don't see any mention of message or custom resources in the goversioninfo docs, so it unfortunately might not be fit for purpose for us. https://github.com/tc-hib/go-winres has facilities to embed custom resources so it may be possible to embed the event message table using that, but the message table would first need to be compiled to the binary format with `windmc`. The message table hasn't changed since it was first committed in 2016, so I imagine it wouldn't be too unreasonable to commit the compiled binary message file into the repo.

File: cmd/docker-proxy/bin_windows_386.go (hunk 2 above)
Comment (TBBle) [reply cmd/docker-proxy/bin_windows_386.go]: Yeah, I feel that a "poison" .syso is a worse-than-doing-nothing solution from a UX perspective.

I'm unsure about stub sysos. That's what hcsshim does for its test suite currently, but in that case there wasn't a go-native solution available, they were hand-created once using the same MSVC tools that you're replacing in the Docker pipeline here. And there's a TODO in the code to fix that anyway, although it wasn't mentioned if the fix is removing the syso and making it dynamic, or just having repeatable tooling to update the .syso when necessary.

I don't think this is a huge issue anyway, `go build` outside the Dockerfile isn't a super-supported workflow on Windows. I've used it on rare occasion, but I had to work out the autogen stuff for myself and hunt down various necessary binaries, so this is a big step forward compared to that in terms of clarify of tooling.

So I reckon what we have now here is fine.

As another (idle) thought, is `goversioninfo` runnable using `go run`? That's how one uses, e.g. [`mksyscall_windows`](https://github.com/microsoft/go-winio/blob/master/syscall.go), and then it gets built/installed automatically at `go generate` time, which means the person building their local dockerd.exe or unit tests doesn't need to separately deal with it.

`go run` might not work well with the build system in this project though, as it wouldn't be able to use the pre-built goversioninfo binary organised in a previous layer, AFAIK. I'd use `go run` if I was working in a simple Go-tooling-only build environment though, sometimes [through a wrapper if the tool itself is not `go run`-able](https://github.com/TBBle/hcsshim/commit/1803a218685ea1c449e1383639dbb459b1473fc7#diff-41f96ff92ea57c6d9e30f148d368d27f2a602885bf2acef714efb544d20ba668).

**Comment by crazy-max** [comment]
> The messages defined in event_messages.mc need to be included in the resources for dockerd.exe in order to support logging to the Windows Event log: [57aef3b](https://github.com/moby/moby/commit/57aef3b49025aac6bb084491478eb461b14b9109). I don't see any mention of message or custom resources in the goversioninfo docs, so it unfortunately might not be fit for purpose for us. https://github.com/tc-hib/go-winres has facilities to embed custom resources so it may be possible to embed the event message table using that, but the message table would first need to be compiled to the binary format with `windmc`. The message table hasn't changed since it was first committed in 2016, so I imagine it wouldn't be too unreasonable to commit the compiled binary message file into the repo.

Good point about the events. I think instead we could use the Manifest-based provider to publish the exact same event schema available in `event_messages.mc`. Not sure if an `instrumentationManifest` is supported by `goversioninfo` though, I will take a look. https://docs.microsoft.com/en-us/windows/win32/etw/publishing-your-event-schema-for-a-manifest-base-provider

File: hack/make/.go-autogen [hunk 9]
```diff
@@ -18,45 +16,12 @@ LDFLAGS="${LDFLAGS} \
...
-
-	makeres dockerd.rc pe-x86-64 autogen/winresources/dockerd/rsrc_amd64.syso
-
-	rm -r autogen/winresources/tmp
+  if [ ! -x "$(command -v goversioninfo)" ]; then
+    >&2 echo "goversioninfo not found, skipping manifesting binary"
+  else
+    (
+      . hack/make/.mkversioninfo
+      go generate -v "${GO_PACKAGE}"
```
[the remark this answers is not in the document]
Comment (crazy-max) [reply hack/make/.go-autogen:24]: > what value is being added by running the `goversioninfo` command indirectly through `go generate`?

Was just to avoid too much scripting and also share the same logic with `go generate` between `.mkversioninfo` and `.go-autogen.ps1`. I we decide to only rely on cross compilation in the future then yes it makes sense to remove the `go generate` logic and have everything in `.mkversioninfo`. See https://github.com/crazy-max/moby/commit/3c5a037b329924251efb2112eb53a66c64d41350

**Comment by crazy-max** [comment]
I have extracted win resources from [`dockerd.exe` 20.10.14](https://download.docker.com/win/static/stable/x86_64/docker-20.10.14.zip) using `go-winres extract`: [winres.zip](https://github.com/moby/moby/files/8442172/winres.zip)

```json
{
  "RT_GROUP_ICON": {
    "#1": {
      "0409": "#1_0409.ico"
    }
  },
  "RT_MANIFEST": {
    "#1": {
      "0409": {
        "identity": {},
        "description": "Docker",
        "minimum-os": "vista",
        "execution-level": "",
        "ui-access": false,
        "auto-elevate": false,
        "dpi-awareness": "unaware",
        "disable-theming": false,
        "disable-window-filtering": false,
        "high-resolution-scrolling-aware": false,
        "ultra-high-resolution-scrolling-aware": false,
        "long-path-aware": false,
        "printer-driver-isolation": false,
        "gdi-scaling": false,
        "segment-heap": false,
        "use-common-controls-v6": false
      }
    }
  },
  "RT_MESSAGETABLE": {
    "#1": {
      "0409": "RT_MESSAGETABLE_#1_0409.bin"
    }
  },
  "RT_VERSION": {
    "#1": {
      "0409": {
        "fixed": {
          "file_version": "20.10.14.0",
          "product_version": "20.10.14.0",
          "type": "Unknown"
        },
        "info": {
          "0000": {
            "FileVersion": "20.10.14",
            "OriginalFileName": "87a90dc",
            "ProductName": "Docker Engine",
            "ProductVersion": "20.10.14"
          }
        }
      }
    }
  }
}
```

Looking at the message table in `RT_MESSAGETABLE_#1_0409.bin` it looks like it supports the events:

```
1 MESSAGETABLE
{
0x1,  		"%1\n"
0x2,  		"debug: %1\n"
0x3,  		"panic: %1\n"
0x4,  		"fatal: %1\n"
0xB,  		"%1 [%2]\n"
0xC,  		"debug: %1 [%2]\n"
0xD,  		"panic: %1 [%2]\n"
0xE,  		"fatal: %1 [%2]\n"
}
```

I guess I can simply add this bin to the working tree so we might not need to use `windmc`.

**Comment by corhere** [comment]
@crazy-max that's a genius way of figuring out the syntax to instruct `go-winres` to add the message table (as a custom resource, in `go-winres` parlance) into the resources! Note that `windmc` was still needed to produce the bin file: [the message table extracted from dockerd.exe 20.10.14 was compiled by `windmc`](https://github.com/moby/moby/blob/63a9e3cc93efabb3759412a11a41a8c07a6e95e5/hack/make/.go-autogen#L54-L57). As `go-winres` does not know how to compile message tables from an editable format, `windmc` would still be needed in the future to recompile the message table if one ever needed to add new entries to it. That's knowledge which could easily be forgotten if the bin file is committed to the repo directly and `windmc` removed from the build scripts, so if we go that route we should keep `event_messages.mc` around and document how to recompile it.

> I think instead we could use the Manifest-based provider to publish the exact same event schema available in `event_messages.mc`.

It doesn't look like that would help. https://docs.microsoft.com/en-us/windows/win32/wes/writing-an-instrumentation-manifest says
> After writing your manifest, use the message compiler to validate the manifest and generate the resource and header files that you include in your provider.

The message compiler it is referring to is `MC.exe`, the Windows SDK tool which GNU binutils `windmc` is a reimplementation of. There's no avoiding the message compiler, it seems.

**Comment by thaJeztah** [comment]
> That's knowledge which could easily be forgotten if the bin file is committed to the repo directly and windmc removed from the build scripts, so if we go that route we should keep event_messages.mc around and document how to recompile it.

💯 agreed; if instructions are needed on how to create this, we should add docs for that or a script (if possible). Not too long ago, I had to do a lot of digging to figure out how some fixtures were created (what options to pass), which was a bit of fun (see https://github.com/moby/moby/pull/42389)) Its ok to be hacky / minimal, as long as it's easy to discover  "how to do this".

**Comment by crazy-max** [comment]
Switch the implementation to use [go-winres](https://github.com/tc-hib/go-winres) which supports message table entries. Also added instructions in case we want to regenerate it.

PTAL @corhere @thaJeztah

**Review by corhere — COMMENTED** [review]
The broad strokes look good to me. There is one issue with the PowerShell script which would cause problems on Windows PowerShell v5.1. And I have some suggestions for your consideration.

File: cmd/docker-proxy/genwinres_windows_386.go [hunk 10]
```diff
@@ -0,0 +1,8 @@
+//go:build windows && 386
+// +build windows,386
+
+//go:generate go-winres make --arch=386 --in=../../cli/winresources/docker-proxy/winres.json --out=../../cli/winresources/docker-proxy/resource.syso
```
[the remark this answers is not in the document]
Comment (crazy-max) [reply cmd/docker-proxy/genwinres_windows_386.go:4]: Oh didn't catch `arch` was an input list!

File: hack/make/.go-autogen.ps1 [hunk 11]
```diff
@@ -27,63 +44,109 @@ Function Get-BuildDateTime() {
...
+}'
 
-    # If you really want to understand this madness below, search the Internet for powershell variables after verbatim arguments... Needed to get double-quotes passed through to the compiler options.
-    # Generate the .syso files containing all the resources and manifest needed to compile the final docker binaries. Both 32 and 64-bit clients.
-    $env:_ag_dockerVersion=$DockerVersion
-    $env:_ag_gitCommit=$CommitString
+    # Write the file
+    $outputFile="$(Get-Location)\cli\winresources\dockerd\winres.json"
+    if (Test-Path $outputFile) { Remove-Item $outputFile }
+    Set-Content -NoNewline -Path $outputFile -Value $mkwinresContents
```
[the remark this answers is not in the document]
Comment (crazy-max) [reply hack/make/.go-autogen.ps1:132]: Yes I used `Set-Content` because I needed `-NoNewline` otherwise `goversioninfo` would fail because of the carriage return `\r\n` in the JSON file. Will se with `go-winres`

File: hack/make/.mkwinres [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+            "ProductName": "${PRODUCT}",
+            "ProductVersion": "${VERSION}",
+            "SpecialBuild": "${GITCOMMIT}"
+          }
+        }
+      }
+    }
+  }
+}
+EOL
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply hack/make/.mkwinres:76]: Do we want to allow building without jq installed? (when building outside of the dockerfile?)

File: hack/make/.go-autogen.ps1 (hunk 11 above)
[the remark this answers is not in the document]
Comment (crazy-max) [reply hack/make/.go-autogen.ps1:132]: Looks good

File: hack/make/.mkwinres (hunk 12 above)
[the remark this answers is not in the document]
Comment (crazy-max) [reply hack/make/.mkwinres:76]: I prefer interpolation changes for this case so we have almost the same behavior between sh/powershell.

**Review by corhere — APPROVED** [review]
LGTM! :shipit:

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by thaJeztah** [comment]
Kicked CI earlier, and all green, except for one flaky test on Windows;

```
=== RUN   TestNetworkDBCRUDTableEntries
    networkdb_test.go:310: Entry existence verification test failed for node2(e7bb4e7b34e3)
2022/04/20 11:21:02 Closing DB instances...
--- FAIL: TestNetworkDBCRUDTableEntries (8.04s)
```

Let's get this one in
